src/dataflow/src/render/join/linear_join.rs

```diff
@@ -77,31 +213,47 @@ where
                 .filter(filter)
                 .project(project);
 
-            // Construct initial join build state.
-            // This state will evolves as we build the join dataflow.
-            let mut join_build_state = JoinBuildState::new(
-                input_mapper.global_columns(*start),
-                &equivalences,
-                &map_filter_project,
+            // Create a join plan based on the linear join.
+            let linear_plan = LinearJoinPlan::create_from(
+                *start,
+                equivalences,
+                order,
+                input_mapper,
+                map_filter_project,
             );
 
-            // This collection will evolve as we join in more inputs.
-            // TODO(mcsherry): determine and apply closure here in `flat_map_ref` form.
-            // TODO(mcsherry): If we plan to use an arrangement, should one exist, then
-            // this is wasteful as it instantiates all rows which are then dropped.
-            let (mut joined, mut errs) = self.collection(&inputs[*start]).unwrap();
-
-            let use_leading_arrangement = start_arr.is_some() && inputs.len() > 1;
-            if !use_leading_arrangement {
-                // NOTE(mcsherry): ideally this code is rarely/never relevant, as the associated logic
-                // could be pushed down to the input and perhaps beyond. I'm not certain under what
-                // circumstance we should just delete it, though.
-
-                // At this point we are able to construct a per-row closure that can be applied once
-                // we have the first wave of columns in place. We will not apply it quite yet, because
-                // we have three code paths that might produce data and it is complicated.
-                let closure = join_build_state.extract_closure();
-                if !closure.is_identity() {
+            // Collect all error streams, and concatenate them at the end.
+            let mut errors = Vec::new();
+
+            // Determine what form the linear spine of updates will take.
+            // This could be a stream, or one of several arrangements.
+            let mut joined = if start_arr.is_some() && inputs.len() > 1 {
+                // Assert that we did not extract an initial closure.
+                assert!(linear_plan.initial_closure.is_none());
```

Comment:
@wangandi I think we can just detect here if `self.arrangement( ... )` exists, rather than take instruction from the join implementation, if that is appropriate. It is what we do for the other inputs, and it may be ok now that we have a better sense for when the first input would be arranged. Do you have thoughts that it might be a bad idea?